src/daemon.cpp

```diff
@@ -0,0 +1,224 @@
+{
+  auto atom_cookie = xcb_intern_atom (c, 0, strlen (atom_name), atom_name);
+  auto atom_reply = std::unique_ptr<xcb_intern_atom_reply_t> (
+      xcb_intern_atom_reply (c, atom_cookie, nullptr));
+
+  // TODO add proper exception
+  auto atom = atom_reply ? atom_reply->atom : throw;
+
+  /* Getting property from atom */
+
+  auto prop_cookie = xcb_get_property (
+      c, 0, root, atom, XCB_GET_PROPERTY_TYPE_ANY, 0, UINT32_MAX);
+  auto prop_reply = std::unique_ptr<xcb_get_property_reply_t> (
+      xcb_get_property_reply (c, prop_cookie, nullptr));
+
+  int prop_length = xcb_get_property_value_length (prop_reply.get ());
+
+  auto *prop = prop_length ? (reinterpret_cast<uint32_t *> (
+                   xcb_get_property_value (prop_reply.get ())))
+                           : throw;
+
+  std::vector<uint32_t> atom_data;
+
+  for (int i = 0; i < prop_length; i++)
+    {
+      atom_data.push_back (prop[i]);
+    }
+
+  return atom_data;
+};
+
+struct monitor_info
+{
+  int x;
+  int y;
+  int width;
+  int height;
+};
+
+std::vector<monitor_info>
+get_monitors ()
+{
+  std::vector<monitor_info> monitors;
+
+  auto screen_resources_reply
+      = std::unique_ptr<xcb_randr_get_screen_resources_current_reply_t> (
+          xcb_randr_get_screen_resources_current_reply (
+              c, xcb_randr_get_screen_resources_current (c, root), nullptr));
+
+  auto *randr_outputs = xcb_randr_get_screen_resources_current_outputs (
+      screen_resources_reply.get ());
+
+  int len = xcb_randr_get_screen_resources_current_outputs_length (
+      screen_resources_reply.get ());
+
+  for (int i = 0; i < len; i++)
+    {
+      auto output_cookie = xcb_randr_get_output_info (c, randr_outputs[i],
+                                                      XCB_TIME_CURRENT_TIME);
+
+      auto output = std::unique_ptr<xcb_randr_get_output_info_reply_t> (
+          xcb_randr_get_output_info_reply (c, output_cookie, nullptr));
+
+      if (output == nullptr || output->crtc == XCB_NONE
+          || output->connection == XCB_RANDR_CONNECTION_DISCONNECTED)
+        {
+          continue;
+        }
+
+      auto crtc = std::unique_ptr<xcb_randr_get_crtc_info_reply_t> (
+          xcb_randr_get_crtc_info_reply (
+              c, xcb_randr_get_crtc_info (c, output->crtc, XCB_CURRENT_TIME),
+              nullptr));
+
+      monitor_info m{ crtc->x, crtc->y, crtc->width, crtc->height };
+      monitors.push_back (m);
+    }
+  return monitors;
+}
+
+struct desktop_info
+{
+  int desktop_number; // XXX Is it necessary?
+  int x;
+  int y;
+  int width;
+  int height;
+  std::string name;
+};
+
+/**
+ * Get an array of desktop_info.
+ *
+ * FIXME Uses EWMH and won't work if WM doesn't support large desktops
+ */
+std::vector<desktop_info>
+get_desktops ()
+{
+  auto monitors = get_monitors ();
+
+  auto number_of_desktops = get_property_value ("_NET_NUMBER_OF_DESKTOPS")[0];
+
+  auto viewport = get_property_value ("_NET_DESKTOP_VIEWPORT");
```

Comment:
I don't think this will work on my os

Reply:
Compositing DEs are not really a target demographics here. You already have you desktop previewer built in